contracts/atomic-swap/src/contract.rs

```diff
@@ -0,0 +1,619 @@
+            }
+            Err(e) => panic!("unexpected error: {:?}", e),
+        }
+
+        // Cannot create, invalid hash
+        let env = mock_env(&sender, &balance);
+        let create = CreateMsg {
+            id: "swap0001".to_string(),
+            hash: "bu115h17".to_string(),
+            recipient: "rcpt0001".into(),
+            end_time: 0,
+            end_height: 123456,
+        };
+        let res = handle(&mut deps, env, HandleMsg::Create(create.clone()));
+        match res {
+            Ok(_) => panic!("expected error"),
+            Err(StdError::GenericErr { msg, .. }) => assert_eq!(
+                msg,
+                "Error parsing hash: Invalid character \'u\' at position 1".to_string()
+            ),
+            Err(e) => panic!("unexpected error: {:?}", e),
+        }
+
+        // Can create, all valid
+        let env = mock_env(&sender, &balance);
+        let create = CreateMsg {
+            id: "swap0001".to_string(),
+            hash: real_hash(),
+            recipient: "rcpt0001".into(),
+            end_time: 0,
+            end_height: 123456,
+        };
+        let res = handle(&mut deps, env, HandleMsg::Create(create.clone())).unwrap();
+        assert_eq!(0, res.messages.len());
+        assert_eq!(log("action", "create"), res.log[0]);
+
+        // Cannot re-create (modify), already existing
+        let new_balance = coins(1, "tokens");
+        let env = mock_env(&sender, &new_balance);
+        let create = CreateMsg {
+            id: "swap0001".to_string(),
+            hash: real_hash(),
+            recipient: "rcpt0001".into(),
+            end_time: 0,
+            end_height: 123456,
+        };
+        let res = handle(&mut deps, env, HandleMsg::Create(create.clone()));
+        match res {
+            Ok(_) => panic!("expected error"),
+            Err(StdError::GenericErr { msg, .. }) => {
+                assert_eq!(msg, "Atomic swap already exists".to_string())
+            }
+            Err(e) => panic!("unexpected error: {:?}", e),
+        }
+    }
+
+    #[test]
+    fn test_release() {
+        let mut deps = mock_dependencies(CANONICAL_LENGTH, &[]);
+
+        let env = mock_env("anyone", &[]);
+        init(&mut deps, env, InitMsg {}).unwrap();
+
+        let sender = HumanAddr::from("sender0001");
+        let balance = coins(1000, "tokens");
+
+        let env = mock_env(&sender, &balance);
+        let create = CreateMsg {
+            id: "swap0001".to_string(),
+            hash: real_hash(),
+            recipient: "rcpt0001".into(),
+            end_time: 0,
+            end_height: 123456,
+        };
+        handle(&mut deps, env.clone(), HandleMsg::Create(create.clone())).unwrap();
+
+        // Cannot release, wrong id
+        let release = HandleMsg::Release {
+            id: "swap0002".to_string(),
+            preimage: preimage(),
+        };
+        let res = handle(&mut deps, env.clone(), release);
```

Comment:
env for release should have no tokens. Use a different `mock_env` like `mock_env(HumanAddr::from("anyone"), &[])`